### PR TITLE
[SGD] Dataset API

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

This PR introduces a minimal dataset api for Ray which is integrated into the TorchTrainer. A dataset takes in an iterable or a parallel iterator, optionally parallelizes some preprocessing transformations, then shards, batches, then trains on the data.  

## Related issue number

N/A

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/util/sgd/data/examples/.gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+ray-project/*
+*.yaml
```
Comment (richardliaw) [inline python/ray/util/sgd/data/examples/.gitignore]: we shouldn't ignore yamls

**Comment by richardliaw** [comment]
Very fun! What does this do exactly? What are the performance behaviors and how do I try it out/does it work out of the box?

File: python/ray/util/sgd/data/imagenet_dataset.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    """
+
+    def __init__(self,
+                 loc,
+                 split,
+                 val_sol_path=None,
+                 max_paths=None,
+                 transform=None):
+        assert split in ["train", "validate"]
+        super(ImageNetDataset, self).__init__(
```
Comment (rkooo567) [inline python/ray/util/sgd/data/imagenet_dataset.py:75]: `super` arguments are not required from Python3 I believe? Does it work when you remove these arguments?

File: python/ray/util/sgd/data/dataset.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    def get_data_loader(self, batch_size=32):
+        materialized = list(self)
+        loader = torch.utils.data.DataLoader(materialized, batch_size)
+        return loader
+
+    def get_sharded_loader(self, n, i, batch_size=32):
+        """Get the ith shard of the dataset assuming it is sharded into
+        n pieces"""
+
+        def converter(x):
```
Comment (maximsmol) [inline python/ray/util/sgd/data/dataset.py:112]: There are three different places where this is defined. This `converter` function, the `__getitem__` method and the `__iter__` method. Maybe we should refactor this logic out?

File: python/ray/util/sgd/data/dataset.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+                if only the driver has the necessary permissions to access
+                the data, remote should be False
+            max_paths: Maximum number of paths to load (passed to `get_paths`)
+
+        """
+        self._paths = self.get_paths(loc, max_paths=max_paths)
+        self.transform = transform
+
+        if remote:
+            data = [_load_remote.remote(self, path) for path in self._paths]
```
Comment (maximsmol) [inline python/ray/util/sgd/data/dataset.py:38]: This remote might end up on a different node (AWS instance) and download the data to that machine. It will then be serialized, put into plasma, and sent over to the local node, effectively re-downloading the data. Not sure if this behavior is intended, but it seems really weird to me. Perhaps we should always force the `_load_remote` onto the current node?

File: python/ray/util/sgd/data/imagenet_dataset.py (hunk 2 above)
Comment (maximsmol) [reply python/ray/util/sgd/data/imagenet_dataset.py:75]: I think you are right too, but I want to note that most of the Ray SGD code currently passes these arguments. In case we decide to use the 3.0 `super()` API we should do that everywhere.

File: python/ray/util/sgd/data/imagenet_dataset.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        tuple: (classes, class_to_idx) where classes are relative to
+    (dir), and class_to_idx is a dictionary.
+
+    Ensures:
+        No class is a subdirectory of another.
+
+    """
+    if sys.version_info >= (3, 5):
+        # Faster and available in Python 3.5 and above
+        classes = [d.name for d in os.scandir(dir) if d.is_dir()]
```
Comment (maximsmol) [inline python/ray/util/sgd/data/imagenet_dataset.py:31]: Nice! Do we support Python <3.5? I don't see any wheels for it.

File: python/ray/util/sgd/data/imagenet_dataset.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        classes = [
+            d for d in os.listdir(dir) if os.path.isdir(os.path.join(dir, d))
+        ]
+    classes.sort()
+    class_to_idx = {classes[i]: i for i in range(len(classes))}
+    return classes, class_to_idx
+
+
+def _load_val_sol(val_sol_path):
+    assert val_sol_path is not None, \
```
Comment (maximsmol) [inline python/ray/util/sgd/data/imagenet_dataset.py:42]: Pretty sure Python will error if arguments without defaults are not provided anyway, so this will never trigger.

File: python/ray/util/sgd/data/examples/.gitignore [hunk 7]
```diff
@@ -0,0 +1 @@
+ray-project/*
```
Comment (maximsmol) [inline python/ray/util/sgd/data/examples/.gitignore:1]: Is this generated by something? I don't see any reference to this folder at all.

File: python/ray/util/sgd/data/utils.py [hunk 8]
```diff
@@ -0,0 +1,9 @@
+def get_arr_partition(arr, n, i):
+    per_part = len(arr) // n
+    extra = len(arr) % n
```
Comment (maximsmol) [inline python/ray/util/sgd/data/utils.py:3]: Having un-even shards might influence training behavior I think. Feel free to correct me here, I'm not certain. If I'm right, it might be important to make using the extra data an explicit choice.

File: python/ray/util/sgd/torch/distributed_torch_runner.py [hunk 9]
```diff
@@ -42,6 +42,7 @@ def setup(self, url, world_rank, world_size):
 
     def _setup_distributed_pytorch(self, url, world_rank, world_size):
         self.world_rank = world_rank
+        self.config["world_rank"] = world_rank
```
Comment (maximsmol) [inline python/ray/util/sgd/torch/distributed_torch_runner.py:45]: We should really come up with a better API for passing system properties around at some point :P Just noting this for the future.

**Review by maximsmol — CHANGES_REQUESTED** [review]
I had some minor comments, overall looks good. However, I don't entirely understand the intended semantics here. Currently, it is not obvious to me whether the code downloads the data or just reads it from disk, whether data is supposed to be available on all instances in its entirety, etc.

Could you clarify how exactly a given instance is supposed to determine what data it should use and how exactly it is supposed to acquire that data?

It seems like the code supports some remote functionality, but the example only uses a local ray cluster. The docstrings in the abstract class mention downloading the data, but the example relies on the data already being present on disk. Sharding happens after all the data has been loaded as well (i.e. all the `load` calls already happened), implying that every agent loads all the data no matter what sharding data it has. I find this confusing, but I am also pretty sure that this is just a matter of better documentation.

That being said, I am very excited about the idea of having a better abstraction for working with distributed datasets from Ray SGD.

File: python/ray/util/sgd/data/dataset.py (hunk 4 above)
Comment (wuisawesome) [reply python/ray/util/sgd/data/dataset.py:38]: This is intentional, there are use cases (like s3) where due to throttling our bandwidth to the datastore is a function of the number of nodes downloading, so we want more nodes to download data and reshuffling is cheap because we have much higher internal bandwidth.

File: python/ray/util/sgd/data/examples/.gitignore (hunk 7 above)
Comment (wuisawesome) [reply python/ray/util/sgd/data/examples/.gitignore:1]: This is generated by anyscale. I'm treating this like an IDE artifact

File: python/ray/util/sgd/data/imagenet_dataset.py (hunk 5 above)
Comment (wuisawesome) [reply python/ray/util/sgd/data/imagenet_dataset.py:31]: good catch

File: python/ray/util/sgd/data/utils.py (hunk 8 above)
Comment (wuisawesome) [reply python/ray/util/sgd/data/utils.py:3]: I don't think shard size should matter because we do a synchronous sgd. This partitioning keeps all partitions within +/- 1 of each other so it should perform well (perhaps batch size alignment of shards could be a little better?). 

Making the shard sizes even would influence training behavior because it would cause non uniform sampling.

File: python/ray/util/sgd/torch/distributed_torch_runner.py (hunk 9 above)
Comment (wuisawesome) [reply python/ray/util/sgd/torch/distributed_torch_runner.py:45]: I think I'll actually be able to rework the code and get rid of this, but I agree it's a bit clunky in general.

**Comment by wuisawesome** [comment]
> I had some minor comments, overall looks good. However, I don't entirely understand the intended semantics here. Currently, it is not obvious to me whether the code downloads the data or just reads it from disk, whether data is supposed to be available on all instances in its entirety, etc.

So the idea is that we should be able to provide a uniform dataset abstraction regardless of where the data originally live. I'll be adding an s3 loader and potentially a few others soon. The data is all in the object store so it in principle it could be accessed from any instance, but the primary intention is for the data to be sharded as transparently as possible. 

> Could you clarify how exactly a given instance is supposed to determine what data it should use and how exactly it is supposed to acquire that data?

The data will be sharded and each instance will get a shard (of approximately equal size). Sharding is meant to be pseudorandom. 

> It seems like the code supports some remote functionality, but the example only uses a local ray cluster. The docstrings in the abstract class mention downloading the data, but the example relies on the data already being present on disk. Sharding happens after all the data has been loaded as well (i.e. all the `load` calls already happened), implying that every agent loads all the data no matter what sharding data it has. I find this confusing, but I am also pretty sure that this is just a matter of better documentation.

In general, we want to support loading data in parallel, but there are some cases where only the head node will be able to read the data (i.e. a local disk, or where permissions are involved, the head typically has more permissions than workers). Perhaps this will be more clear once the s3 loader is added, and maybe I should just put that all into 1 PR. 

> That being said, I am very excited about the idea of having a better abstraction for working with distributed datasets from Ray SGD.

:)

File: python/ray/util/sgd/data/dataset.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+from ray.util.sgd.data.utils import get_arr_partition
+
+
+@ray.remote
+def _load_remote(dataset, path):
+    return dataset.load(path)
+
+
+class Dataset(ABC, torch.utils.data.IterableDataset):
```
Comment (amogkam) [inline python/ray/util/sgd/data/dataset.py:15]: Maybe we should change this to `FileDataset`? I think there's an opportunity here to make a more general `Dataset` class that doesn't necessarily have to load training data from a filepath.

File: python/ray/util/sgd/data/dataset.py (hunk 4 above)
Comment (amogkam) [inline python/ray/util/sgd/data/dataset.py:38]: I think you can just do `_load_remote = ray.remote(load)` and not have to define it outside the class?

File: python/ray/util/sgd/data/dataset.py [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+        self._paths = self.get_paths(loc, max_paths=max_paths)
+        self.transform = transform
+
+        if remote:
+            data = [_load_remote.remote(self, path) for path in self._paths]
+        else:
+            data = [ray.put(self.load(path)) for path in self._paths]
+
+        random.seed(0)
+        random.shuffle(data)
```
Comment (amogkam) [inline python/ray/util/sgd/data/dataset.py:43]: Making a note here: we'd eventually want to use Parallel Iterator's shuffling here.

File: python/ray/util/sgd/data/imagenet_dataset.py (hunk 6 above)
Comment (amogkam) [reply python/ray/util/sgd/data/imagenet_dataset.py:42]: I think this is just checking to make sure `None` isn't passed in as the argument.

**Review by amogkam — COMMENTED** [review]
Left some comments. Overall very good abstractions for training Dataset.

File: python/ray/util/sgd/data/utils.py (hunk 8 above)
Comment (maximsmol) [reply python/ray/util/sgd/data/utils.py:3]: Alright, looks good then :+1:

File: python/ray/util/sgd/data/examples/.gitignore (hunk 7 above)
Comment (maximsmol) [reply python/ray/util/sgd/data/examples/.gitignore:1]: Got it

File: python/ray/util/sgd/data/dataset.py (hunk 4 above)
Comment (maximsmol) [reply python/ray/util/sgd/data/dataset.py:38]: I am interested to see more examples of this, but it makes more sense now, thanks!

File: python/ray/util/sgd/data/dataset.py (hunk 10 above)
Comment (wuisawesome) [reply python/ray/util/sgd/data/dataset.py:15]: so path wasn't intended to strictly be a filepath (it could be an s3 path or hdfs path for example), but perhaps there's a better name...

Comment (amogkam) [reply python/ray/util/sgd/data/dataset.py:15]: Right but this still assumes you're reading files from somewhere. What do you do if you have your training data in memory? I'm thinking we should also have a general `Dataset` that just takes in an iterable.

**Comment by richardliaw** [comment]
Hey; this PR is getting pretty big, and I'm worried that since it touches a bunch of fast moving parts that we're just going to be chasing rebases. 

Can you break this up into the following:

* Dataset + Parallel iterator changes.
  * Include standalone tests.

* Any changes to RaySGD 
  * Include docs
  * Include examples
  * Include integration tests

File: python/ray/util/sgd/torch/torch_trainer.py [hunk 12]
```diff
@@ -379,7 +377,8 @@ def train(self,
               profile=False,
               reduce_results=True,
               max_retries=3,
-              info=None):
+              info=None,
+              dataset=None):
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/torch_trainer.py:381]: docstring

File: python/ray/util/sgd/torch/torch_trainer.py [hunk 13]
```diff
@@ -414,11 +413,13 @@ def train(self,
                 length will be equal to ``num_workers``.
         """
         assert max_retries >= 0, "`max_retries` must be non-negative."
+        assert dataset is not None or self.data_creator, \
+            "Must specify either a data creator or a dataset"
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/torch_trainer.py:417]: Can you typecheck the dataset object?

File: python/ray/util/sgd/torch/torch_runner.py [hunk 14]
```diff
@@ -193,9 +199,18 @@ def train_epoch(self, num_steps=None, profile=False, info=None):
...
+                # need two lists.
+                def format_batch(batch):
+                    features, targets = zip(*batch)
+                    return torch.cat(features), torch.cat(targets)
+
+                iterator = map(format_batch, iterator)
             if num_steps:
-                iterator = itertools.islice(iter(self.train_loader), num_steps)
+                iterator = itertools.islice(iterator, num_steps)
             train_stats = self.training_operator.train_epoch(iterator, info)
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/torch_runner.py:214]: @maximsmol what did you intend to do here?

File: python/ray/util/sgd/tests/test_torch.py [hunk 15]
```diff
@@ -9,6 +9,7 @@
 from torch.utils.data import DataLoader
 
 import ray
+import ray.util.sgd.data.examples.mlp_identity
```
Comment (richardliaw) [inline python/ray/util/sgd/tests/test_torch.py:12]: would it be possibe for you to just import the function you want instead of the entire module?

Comment (wuisawesome) [reply python/ray/util/sgd/tests/test_torch.py:12]: So the functions I want include model_creator and optimizer_creator. How should I got about importing those shadowing the existing names?

`from ray.util.... import model_creator as model_creator_dataset` feels a little funky.

File: python/ray/util/sgd/torch/torch_runner.py (hunk 14 above)
Comment (wuisawesome) [reply python/ray/util/sgd/torch/torch_runner.py:214]: Note that line 203 calls iter on the train_loader.

File: doc/source/raysgd/raysgd_ref.rst [hunk 16]
```diff
@@ -33,3 +33,12 @@ TFTrainer
     :members:
 
     .. automethod:: __init__
+
+Dataset
+-------
+
+.. autoclass:: ray.util.sgd.tf.TFTrainer
```
Comment (richardliaw) [inline doc/source/raysgd/raysgd_ref.rst:40]: this is wrong?

File: python/ray/util/sgd/data/dataset.py [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+class Dataset():
+    """A simple Dataset abstraction for RaySGD.
+
+    This dataset is designed to work with RaySGD trainers (currently just
+    Torch) to provide support for streaming large external datasets, and built
+    in sharding.
+
+    .. code-block:: python
+
+            def to_mat(x):
```
Comment (richardliaw) [inline python/ray/util/sgd/data/dataset.py:13]: I think this has an extra indent

File: python/ray/util/sgd/torch/torch_trainer.py [hunk 18]
```diff
@@ -473,7 +493,8 @@ def _train_epoch(self, num_steps=None, profile=False, info=None):
 
         success = check_for_failure(remote_worker_stats)
         if success:
-            return success, [local_worker_stats] + ray.get(remote_worker_stats)
+            a, b = success, [local_worker_stats] + ray.get(remote_worker_stats)
+            return a, b
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/torch_trainer.py:497]: any reason for this?

**Review by richardliaw — COMMENTED** [review]
OK generally looks good. Let a couple comments for docs.

**Review by richardliaw — APPROVED** [review]

File: python/ray/util/sgd/data/dataset.py [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+
+    def __init__(self,
+                 data,
+                 batch_size=32,
+                 download_func=None,
+                 max_concurrency=0,
+                 transform=None):
+        par_iter = None
+        if isinstance(data, ParallelIterator):
+            par_iter = data.repartition(1)
```
Comment (amogkam) [inline python/ray/util/sgd/data/dataset.py:62]: Is there a reason we have to repartition to 1 here instead of leaving as is?

Comment (wuisawesome) [reply python/ray/util/sgd/data/dataset.py:62]: Original rationale was so that max_concurrency would set the cluster-wide parallelism. Do you think we should just take it out?

Comment (maximsmol) [reply python/ray/util/sgd/data/dataset.py:62]: I think the original design with forcing exactly `max_concurrency` shards makes more sense. Is there a compelling reason (maybe performance-wise or where in the cluster shards are placed) to use pre-sharded iterators?

File: python/ray/util/sgd/tests/test_torch.py [hunk 20]
```diff
@@ -591,34 +633,60 @@ def test_wrap_ddp(ray_start_2_cpus, tmp_path):  # noqa: F811
...
+                return False, None
+            if "NCCL" in err.args[0]:  # there is no specific error message
+                return False, None
+
+            raise err
 
         success = check_for_failure(remote_worker_stats)
         if success:
-            return success, [local_worker_stats] + ray.get(remote_worker_stats)
+            a, b = success, [local_worker_stats] + ray.get(remote_worker_stats)
```
Comment (maximsmol) [inline python/ray/util/sgd/tests/test_torch.py:672]: Why was this changed lol

File: python/ray/util/sgd/torch/torch_runner.py (hunk 14 above)
Comment (maximsmol) [reply python/ray/util/sgd/torch/torch_runner.py:214]: @richardliaw are you talking about my "downstream state" API proposal?

**Review by maximsmol — COMMENTED** [review]
Good stuff. Left a few small comments, mostly just questions. Love the failure tolerance test changes. The API looks very clean.

File: python/ray/util/sgd/data/dataset.py (hunk 19 above)
Comment (wuisawesome) [reply python/ray/util/sgd/data/dataset.py:62]: There are some use cases where this could lead to a lot of data shuffling, but it wouldn't be a very good use of parallel iterator to begin with. 

I propose we leave the repartition here, then remove/rethink the signature if there's a compelling performance reason.

File: doc/source/raysgd/raysgd_dataset.rst [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+distributed data.
+
+.. tip:: Get in touch with us if you're using or considering using `RaySGD <https://forms.gle/26EMwdahdgm7Lscy9>`_!
+
+Setting up a dataset
+--------------------
+
+A dataset can be constructed via any python iterable, or a ``ParallelIterator``. Optionally, a batch size, download function, concurrency, and a transformation can also be specified.
+
+When constructing a dataset, a download function can be specified. For example, if a dataset is initialized with a set of paths, a download function can be specified which converts those paths to ``(input, label)`` tuples. The download function can be executed in parallel via ``max_concurrency``. This may be useful if the backing datastore has rate limits, there is high overhead associated with a download, or downloading is computationally expensive.
```
[the remark this answers is not in the document]
Comment (wuisawesome) [reply doc/source/raysgd/raysgd_dataset.rst:14]: I'm not exactly sure what you mean, I think all of those things fall under the intended purpose.

File: python/ray/util/sgd/data/dataset.py (hunk 19 above)
Comment (amogkam) [reply python/ray/util/sgd/data/dataset.py:62]: OK that sounds good
